repartitions/algorithm.py

```diff
@@ -0,0 +1,200 @@
+            for uc in fixed_user_campaigns:
+                repartition[proposition_id_to_index[uc.fixed_to.id]].append(uc)
+
+        if len(user_campaigns) > 0:
+            places = []
+            s = 0  # Nombre de places allouées pour le groupe
+            for i, proposition in enumerate(propositions):
+                n = -already_used[i]
+                n -= len(repartition)
+                n -= penalty[i]
+                s += n
+                places.append(n)
+
+            while s < len(user_campaigns):
+                for i in range(len(places)):
+                    s += 1
+                    places[i] += 1
+
+            cost_matrix = []
+            for uc in user_campaigns:
+                cost_matrix.append(generate_line(uc, propositions, places))
+
+            # make the matrix square
+            while len(cost_matrix) < len(cost_matrix[0]):
+                cost_matrix.append([0] * len(cost_matrix[0]))
+
+            links = linear_sum_assignment(cost_matrix)
+
+            for user_index, place_index in zip(links[0], links[1]):
+                if user_index >= len(user_campaigns):
+                    break
+
+                project_index = get_project_index(place_index, places)
+                uc = user_campaigns[user_index]
+                repartition[project_index].append(uc)
+
+            repartition_size = list(map(len, repartition))
+            repartition_cardinal_diff = max(repartition_size) - min(repartition_size)
+
+            penalty[np.argmax(repartition_size)] += 1
+        else:
+            repartition_cardinal_diff = 0  # avoid being stuck in an infinite loop
+
+    print_repartition = []
+    for i in repartition:
+        print_repartition.append(len(i))
+
+    return repartition
+
+
+# Checks the repartition is always even
+def make_reparitition_proxy(
+    category: Category, propositions: List[Proposition], already_used: List[int]
+) -> List[List[User]]:
+    group_cardinal_diff = 2
+    groups = []
+
+    already_used = np.array(already_used)
+    penalty = np.array([0] * len(already_used))
+    while group_cardinal_diff > 1:
+        used_places = already_used + penalty
+        groups = make_repartition_for_category(category, propositions, used_places)
+
+        rep = [0] * len(used_places)
+        for i in range(len(propositions)):
+            rep[i] = len(groups[i])
+        group_cardinal = np.array(rep) + already_used
+        penalty[np.argmax(group_cardinal)] += 1
+        group_cardinal_diff = max(group_cardinal) - min(group_cardinal)
+
+    return groups
+
+
+def make_reparition(campaign: Campaign) -> List[Group]:
+    propositions = list(campaign.propositions.all())
+    categories = list(campaign.categories.all())
+
+    already_used = [0] * len(propositions)
+    store = {}
+
+    for i in range(len(categories)):
+        category = categories[i]
+
+        if category.users_campaign.count() == 0:
+            continue
+
+        # Check the repartition is feasible
```

Comment:
Make it a function then?

Reply:
Not necessarily, this is a pretty standard pattern